Added basic pydantic models;

xoryouyou opened the pull request:
Changed scraper to use new models;
changed old .txt files to proper json base on pydantic models

fvsamson commented:
#### Mini review
* In [`scraper.py`, line 25](https://github.com/xoryouyou/bsi-technical-guidelines-markdown/pull/4/files/c9de295269247a06c3e8771e17332a7ec4da6818..0e366f6231f6398d3a9d0a5d60ef166cb9c4ebfb#diff-fbb4c543dc45944cc8e497ec8bca39c82308d65946a1fecae30bf7cc72151112R25): Is the missing, concluding "**t**" of the second text string an oversight / copy&paste error on your side or the IT-GS people's?
* [Nitpick] Comment in [`scraper.py`, line 332](https://github.com/xoryouyou/bsi-technical-guidelines-markdown/pull/4/files/c9de295269247a06c3e8771e17332a7ec4da6818..0e366f6231f6398d3a9d0a5d60ef166cb9c4ebfb#diff-fbb4c543dc45944cc8e497ec8bca39c82308d65946a1fecae30bf7cc72151112R332): IMO it is a "download directory" (to download data to), not a "download**s** directory" (to amass downloads in).  Just my English language gut feeling (while often right, it sometimes misleads me grossly) as a proficient, but non-native speaker.  Well, never mind if your mileage varies.

BTW, I am bad at Python (and most other "modern" programming languages), so please do not rely on my review(s).

fvsamson reviewed (commented):
Much better, this PR now constitutes a vast improvement over the original state!

<sub>Ultimately, you may be able to make me overcome my aversion against JSON ("no comments"!) and make me learn Python (which I intend to do for more than 15 years). :smile:</sub>

xoryouyou commented:
Thanks for all the input!

I've fixed a few minor things and moved the data "storage" to a `repository.json`

Although not perfect it is now in a semi workable state to have a strict validator/schema library like pydantic handle the de-/serialization.

This allows for easy future transfer to a proper database and keeping a consistend format.

Also I've restructured a few code sections and got rid of all the duplication according to sonarqube.

The `__init__.py` is a python thing to tell the interpreter that "this directory contains python files to look at" see https://docs.python.org/3/reference/import.html#regular-packages

Bottom line is: A new basis for future improvements 😄

fvsamson commented:
BTW, the branch `datastructuring` (which has been merged into `main` by this PR) still exists.  In order not to forget deleting feature branches after merging them into a repo's default branch, I usually enable the option to automatically delete branches after merging in GitHub's repository settings: One can still restore ("undelete") an automatically deleted branch for weeks, but ultimately it is really gone for good (as intended).

fvsamson commented:
> Thanks for all the input!

My pleasure! :smiley:

BTW, sorry for these late replies: Tonight, for the first time in the past weeks, I have time *and* motivation to do more than just glancing over your recent changes here.

> I've fixed a few minor things and moved the data "storage" to a `repository.json`

I am not yet convinced that [commit `6c6ff9c` "merged all into one repository.json"](https://github.com/xoryouyou/bsi-technical-guidelines-markdown/pull/4/commits/6c6ff9c06f30b477a86c8a32eec746c0ba15a707#diff-9ce0788e832e70d86d53c78c795bb32d50752d101d57044a04094db32abcb870) was a good move.  While both (IT-GS & BSI TRs) are published by BSI, they dwell in separate (AFAICS currently "fully separate"; more on that below) realms; they also are generated by vastly different processes: IT-GS by one (actually, since recently: two) teams in a (hopefully) coherent manner (GS file names definitely adhere to a single, specific format, from which additional information can be obtained for grouping documents etc.), but BSI TRs are potentially created by any team at BSI (outside of BSI's own administration) to their liking (content, its depth, its breadth, the document length, its structure etc.), only TR numbers and an empty ODT / DOCX template file for TRs are centrally maintained.  Additionally GS might undergo further structural changes; e.g. its structure and look was altered much in the past decade and it is still not really well machine-processable.

In addition to these basic, structural considerations, I can imagine cases in which two separate "data storages" for IT-GS and BSI TRs enhance robustness, both when filling them (by the scraper) and later when using them.  Imagine a scraped IT-GS page contains a link to a TR (either the PDF or its landing page; AFAICS such link(s) do not exist yet, and while I know that the GS teams are a bit reluctant to reference anything not created by them except for ISO 2700x, I also assume they will reference BSI TRs when really necessary or requested by superiors): When that TR link is being scraped using a separate "GS data storage", it clearly falls into the category accompanying documents and will be handled as such when stored (or not) and retrieved / evaluated.  I can imagine (without having looked at the current code, hence this is a wild guess) that logic and implementation weaknesses may result in mishaps (when using a single "data storage") which can be avoided by using two separate "data storages".  Ultimately this point is a bit constructed, as the single JSON file / data-structure is (on its first level) subdivided into a `trs` and [a `grundschutz_bausteine` section](https://github.com/xoryouyou/bsi-technical-guidelines-markdown/pull/4/commits/6c6ff9c06f30b477a86c8a32eec746c0ba15a707#diff-9ce0788e832e70d86d53c78c795bb32d50752d101d57044a04094db32abcb870R3448), still … see next paragraph.

Furthermore, the ultimate goal is to auto-generate two Markdown files each with some introductory text followed by a big table with each row containing metadata for single document with embedded links to the corresponding PDF and its Markdown conversion, one for all BSI TRs and one for the whole IT-GS; at least that is what I gathered from our discussion in the comment section of PR #1.  Do you see any drawback by introducing this separation at the earliest point of time (i.e. when scraping) instead of later?  Note that these considerations are only about the collected data, this still can be achieved by the codebase you beautifully unified lately.

Last but not least, two separate "data storages" provide the freedom to differ in the format of an record (i.e. its data fields).
<sub>Although you found a common set of data fields which fit well, despite GS and TRs not having anything in common, and I do believe that the data collection ("scraping") and the data evaluation / transformation (i.e. generate Markdown) phases should be kept clearly separated, hence deriving e.g. additional information (for grouping documents etc.) from the originally downloaded and recorded data shall not happen in the data collecting step.</sub>

OTOH, it is just software and things can always be altered later, though the labour costs for fundamental changes usually rise exponentially during project development.

I well understand the idea that by creating a single "data storage" in JSON format now, it is easier to later switch to a DBMS for "data storage" by storing the individual datasets (which are currently recorded in a single JSON file) in database records.  But in general, this consideration assumes that DBMS instances are costly (CPU, RAM, storage, and for non-OSS DBMSes also financially) and hence one should minimise the number of them running concurrently (down to a single one, at best), which is no longer true in times of `sqlite` being ubiquitous.  Well, I have no idea what you specifically had in mind when considering this here at GitHub, but the Ubuntu-based CI-runner instances at GitHub comprise a (virtualised,) basic Ubuntu installation which sure includes `sqlite` (and other OSS DBMSes can be easily installed by using `apt` or `pkcon` in a shell script section of your CI configuration YAML file at GitHub).

**But please mind:** I am just brainstorming (late in the evening, in my private time), showering you with ideas which may or may not be valid, reasonable etc.  This is your baby, you are the proficient Python coder (not me!), I can only claim to have aquired broad IT and computer science experience and knowledge, though did not have time and energy to read into and try out some "modern IT technologies" of the last two decades (but have for many others, e.g. `git`, GitHub & Gitlab & Forgejo / Codeberg, GitHub actions and CI workflows).

> Although not perfect it is now in a semi workable state to have a strict validator/schema library like pydantic handle the de-/serialization.

:rocket:
 
> This allows for easy future transfer to a proper database and keeping a consistent format.

Well, see above.
<sub>*Maybe … two databases?* :stuck_out_tongue_winking_eye:</sub>

> Also I've restructured a few code sections and got rid of all the duplication according to sonarqube.

<sup>Ooops, thank you very much!  I was no aware of Sonar(Qube), which surprises me looking at its properties and capabilities.  But then, I have not worked in the field of ensuring code safety and security (usually by static and / or dynamic code analysis) for more than 15 years; still I assume I should at least have read about it at LWN.net etc., but actually never did (or forgot that I did).

> The `__init__.py` is a python thing to tell the interpreter that "this directory contains python files to look at" see https://docs.python.org/3/reference/import.html#regular-packages

Thanks, now I know why I have seen it before, elsewhere.
 
> Bottom line is: A new basis for future improvements 😄

LGTM!
